### PR TITLE
[flang][cuda] Extends matching distance computation

### PR DESCRIPTION
**Pull request by clementval**
Extends the computation of the matching distance in the generic resolution to support options described in the table: https://docs.nvidia.com/hpc-sdk/archive/24.3/compilers/cuda-fortran-prog-guide/index.html#cfref-var-attr-unified-data

Options are added as language features in the `SemanticsContext` and a flag is added in bbc for testing purpose.

### CODE REVIEW

File: flang/include/flang/Common/Fortran-features.h [hunk 1]
```diff
@@ -49,7 +49,7 @@ ENUM_CLASS(LanguageFeature, BackslashEscapes, OldDebugLines,
     IndistinguishableSpecifics, SubroutineAndFunctionSpecifics,
     EmptySequenceType, NonSequenceCrayPointee, BranchIntoConstruct,
     BadBranchTarget, ConvertedArgument, HollerithPolymorphic, ListDirectedSize,
-    NonBindCInteroperability)
+    NonBindCInteroperability, GpuManaged, GpuUnified)
```
Comment (klausler) [inline flang/include/flang/Common/Fortran-features.h:52]: should these names be `CUDA` rather than `GPU`?

File: flang/lib/Common/Fortran.cpp [hunk 2]
```diff
@@ -102,7 +103,13 @@ std::string AsFortran(IgnoreTKRSet tkr) {
 /// dummy argument attribute while `y` represents the actual argument attribute.
 bool AreCompatibleCUDADataAttrs(std::optional<CUDADataAttr> x,
     std::optional<CUDADataAttr> y, IgnoreTKRSet ignoreTKR,
-    bool allowUnifiedMatchingRule) {
+    bool allowUnifiedMatchingRule, const LanguageFeatureControl *features) {
+  bool isGpuManaged = features
```
Comment (klausler) [inline flang/lib/Common/Fortran.cpp:107]: Please use braced initialization in semantics.

File: flang/lib/Semantics/expression.cpp [hunk 3]
```diff
@@ -2501,8 +2501,13 @@ static constexpr int cudaInfMatchingValue{std::numeric_limits<int>::max()};
 
 // Compute the matching distance as described in section 3.2.3 of the CUDA
 // Fortran references.
-static int GetMatchingDistance(const characteristics::DummyArgument &dummy,
+static int GetMatchingDistance(const common::LanguageFeatureControl &features,
+    const characteristics::DummyArgument &dummy,
     const std::optional<ActualArgument> &actual) {
+  bool isGpuManaged = features.IsEnabled(common::LanguageFeature::GpuManaged);
```
Comment (klausler) [inline flang/lib/Semantics/expression.cpp:2507]: Please use braced initialization in semantics.

File: flang/lib/Semantics/expression.cpp [hunk 4]
```diff
@@ -2501,8 +2501,13 @@ static constexpr int cudaInfMatchingValue{std::numeric_limits<int>::max()};
...
+  bool isGpuUnified = features.IsEnabled(common::LanguageFeature::GpuUnified);
+  // assert((isGpuManaged != isGpuUnified) && "expect only one enabled.");
```
Comment (klausler) [inline flang/lib/Semantics/expression.cpp:2509]: Either delete or change to a `CHECK()`.

File: flang/lib/Semantics/expression.cpp [hunk 5]
```diff
@@ -2546,15 +2557,33 @@ static int GetMatchingDistance(const characteristics::DummyArgument &dummy,
       return 2;
     }
   } else if (*dummyDataAttr == common::CUDADataAttr::Managed) {
-    if (!actualDataAttr || *actualDataAttr == common::CUDADataAttr::Device) {
+    if (!actualDataAttr) {
+      if (isGpuUnified) {
```
Comment (klausler) [inline flang/lib/Semantics/expression.cpp:2561]: Would be more readable as `return isGpuUnified ? 1 : isGpuManaged ? 0 : cudaInfMatchingValue;`.

File: flang/lib/Semantics/expression.cpp [hunk 6]
```diff
@@ -2546,15 +2557,33 @@ static int GetMatchingDistance(const characteristics::DummyArgument &dummy,
...
       return cudaInfMatchingValue;
     } else if (*actualDataAttr == common::CUDADataAttr::Managed) {
       return 0;
     } else if (*actualDataAttr == common::CUDADataAttr::Unified) {
       return 1;
     }
   } else if (*dummyDataAttr == common::CUDADataAttr::Unified) {
-    if (!actualDataAttr || *actualDataAttr == common::CUDADataAttr::Device) {
+    if (!actualDataAttr) {
+      if (isGpuUnified) {
```
Comment (klausler) [inline flang/lib/Semantics/expression.cpp:2578]: same here

File: flang/include/flang/Common/Fortran-features.h (hunk 1 above)
Comment (clementval) [reply flang/include/flang/Common/Fortran-features.h:52]: Yeah let's use CUDA until we have other use cases.

File: flang/lib/Semantics/expression.cpp (hunk 4 above)
Comment (clementval) [reply flang/lib/Semantics/expression.cpp]: Updated

File: flang/lib/Semantics/expression.cpp (hunk 5 above)
Comment (clementval) [reply flang/lib/Semantics/expression.cpp]: Yeah I like it more like that. Updated

File: flang/lib/Common/Fortran.cpp [hunk 7]
```diff
@@ -102,7 +103,13 @@ std::string AsFortran(IgnoreTKRSet tkr) {
 /// dummy argument attribute while `y` represents the actual argument attribute.
 bool AreCompatibleCUDADataAttrs(std::optional<CUDADataAttr> x,
     std::optional<CUDADataAttr> y, IgnoreTKRSet ignoreTKR,
-    bool allowUnifiedMatchingRule) {
+    bool allowUnifiedMatchingRule, const LanguageFeatureControl *features) {
+  bool isCudaManaged = features
```
Comment (klausler) [inline flang/lib/Common/Fortran.cpp:107]: Braced initialization, please.

Comment (clementval) [reply flang/lib/Common/Fortran.cpp:107]: Sorry :-(

**Review by klausler — APPROVED** [review]

**Pull request merged**